libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -32,85 +32,363 @@ using namespace solidity;
 using namespace solidity::frontend;
 using namespace solidity::langutil;
 
-void ConstantEvaluator::endVisit(UnaryOperation const& _operation)
+using TypedRational = ConstantEvaluator::TypedRational;
+
+namespace
+{
+
+/// Check whether (_base ** _exp) fits into 4096 bits.
+bool fitsPrecisionExp(bigint const& _base, bigint const& _exp)
 {
-	auto sub = type(_operation.subExpression());
-	if (sub)
-		setType(_operation, sub->unaryOperatorResult(_operation.getOperator()));
+	if (_base == 0)
+		return true;
+
+	solAssert(_base > 0, "");
+
+	size_t const bitsMax = 4096;
+
+	unsigned mostSignificantBaseBit = boost::multiprecision::msb(_base);
+	if (mostSignificantBaseBit == 0) // _base == 1
+		return true;
+	if (mostSignificantBaseBit > bitsMax) // _base >= 2 ^ 4096
+		return false;
+
+	bigint bitsNeeded = _exp * (mostSignificantBaseBit + 1);
+
+	return bitsNeeded <= bitsMax;
 }
 
-void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
+/// Checks whether _mantissa * (2 ** _expBase10) fits into 4096 bits.
+bool fitsPrecisionBase2(bigint const& _mantissa, uint32_t _expBase2)
 {
-	auto left = type(_operation.leftExpression());
-	auto right = type(_operation.rightExpression());
-	if (left && right)
+	return fitsPrecisionBaseX(_mantissa, 1.0, _expBase2);
+}
+
+}
+
+optional<rational> ConstantEvaluator::evaluateBinaryOperator(Token _operator, rational const& _left, rational const& _right)
+{
+	bool fractional = _left.denominator() != 1 || _right.denominator() != 1;
+	switch (_operator)
 	{
-		TypePointer commonType = left->binaryOperatorResult(_operation.getOperator(), right);
-		if (!commonType)
-			m_errorReporter.fatalTypeError(
-				6020_error,
-				_operation.location(),
-				"Operator " +
-				string(TokenTraits::toString(_operation.getOperator())) +
-				" not compatible with types " +
-				left->toString() +
-				" and " +
-				right->toString()
-			);
-		setType(
-			_operation,
-			TokenTraits::isCompareOp(_operation.getOperator()) ?
-			TypeProvider::boolean() :
-			commonType
-		);
+	//bit operations will only be enabled for integers and fixed types that resemble integers
+	case Token::BitOr:
+		if (fractional)
+			return nullopt;
+		else
+			return _left.numerator() | _right.numerator();
+	case Token::BitXor:
+		if (fractional)
+			return nullopt;
+		else
+			return _left.numerator() ^ _right.numerator();
+	case Token::BitAnd:
+		if (fractional)
+			return nullopt;
+		else
+			return _left.numerator() & _right.numerator();
+	case Token::Add: return _left + _right;
+	case Token::Sub: return _left - _right;
+	case Token::Mul: return _left * _right;
+	case Token::Div:
+		if (_right == rational(0))
+			return nullopt;
+		else
+			return _left / _right;
+	case Token::Mod:
+		if (_right == rational(0))
+			return nullopt;
+		else if (fractional)
+		{
+			rational tempValue = _left / _right;
+			return _left - (tempValue.numerator() / tempValue.denominator()) * _right;
```

Comment:
Does this also work for negative `left`?

Reply:
Has been moved from https://github.com/ethereum/solidity/pull/10321/files#diff-5cc133bfc337e9ca7fe7ee9842d5ddfa492fa559149cf8acec4a7d9ede50c9aaL1109 but I will add a test.